File: web/src/lib/confetti.ts
/**
 * Konfetti ohne Dependency: ein Canvas über allem, ein paar Dutzend
 * Papierschnipsel mit Schwerkraft, danach räumt es sich selbst wieder ab.
 *
 * Bewusst klein gehalten — die App hat außer React keine Abhängigkeiten, und
 * dafür lohnt keine Bibliothek. Wer `prefers-reduced-motion` gesetzt hat,
 * bekommt gar nichts: die Feier ist Zugabe, nie Information.
 */

/** Standardfarben: Tape, Kreide und die Grade-Skala. */
const DEFAULT_VARS = [
  '--color-tape',
  '--color-grade-yellow',
  '--color-grade-green',
  '--color-chalk',
  '--color-grade-purple',
];

export interface BurstOptions {
  /** Ursprung in Viewport-Anteilen (0..1). Ohne Angabe: Mitte, etwas erhöht. */
  x?: number;
  y?: number;
  /** Anzahl Schnipsel. */
  count?: number;
  /** Farben als CSS-Werte; ohne Angabe die Palette der App. */
  colors?: string[];
  /** Anfangsgeschwindigkeit — größer heißt weiter. */
  power?: number;
  /** Öffnungswinkel nach oben, in Radiant. */
  spread?: number;
}

interface Particle {
  x: number;
  y: number;
  vx: number;
  vy: number;
  w: number;
  h: number;
  color: string;
  spin: number;
  angle: number;
  life: number;
  age: number;
}

const GRAVITY = 0.32;
const DRAG = 0.987;

let canvas: HTMLCanvasElement | null = null;
let ctx: CanvasRenderingContext2D | null = null;
let particles: Particle[] = [];
let frame = 0;

function reducedMotion(): boolean {
  return window.matchMedia?.('(prefers-reduced-motion: reduce)').matches ?? false;
}

/** `var(--color-tape)` → `#e4572e`; Canvas kennt keine CSS-Variablen. */
function resolve(color: string): string {
  const name = /^var\((--[\w-]+)\)$/.exec(color.trim())?.[1];
  if (!name) return color;
  return getComputedStyle(document.documentElement).getPropertyValue(name).trim() || '#e4572e';
}

function palette(): string[] {
  return DEFAULT_VARS.map((v) => resolve(`var(${v})`));
}

function ensureCanvas(): CanvasRenderingContext2D | null {
  if (!canvas) {
    canvas = document.createElement('canvas');
    canvas.setAttribute('aria-hidden', 'true');
    // Über allem, auch über dem Feier-Dialog: dessen abgedunkelter, weich
    // gezeichneter Hintergrund schluckt sonst genau das Konfetti, das zur
    // Feier gehört. Klicks gehen weiter durch (pointer-events:none).
    canvas.style.cssText =
      'position:fixed;inset:0;pointer-events:none;z-index:60;width:100%;height:100%';
    document.body.appendChild(canvas);
    ctx = canvas.getContext('2d');
  }
  const dpr = Math.min(window.devicePixelRatio || 1, 2);
  const w = window.innerWidth;
  const h = window.innerHeight;
  if (canvas.width !== w * dpr || canvas.height !== h * dpr) {
    canvas.width = w * dpr;
    canvas.height = h * dpr;
    ctx?.setTransform(dpr, 0, 0, dpr, 0, 0);
  }
  return ctx;
}

function teardown(): void {
  if (frame) cancelAnimationFrame(frame);
  frame = 0;
  particles = [];
  canvas?.remove();
  canvas = null;
  ctx = null;
}

function tick(): void {
  const c = ctx;
  if (!c || !canvas) return teardown();
  c.clearRect(0, 0, window.innerWidth, window.innerHeight);

  for (const p of particles) {
    p.age++;
    p.vy += GRAVITY;
    p.vx *= DRAG;
    p.vy *= DRAG;
    p.x += p.vx;
    p.y += p.vy;
    p.angle += p.spin;

    // Am Ende ausblenden, statt die Schnipsel hart verschwinden zu lassen.
    const fade = Math.max(0, Math.min(1, (p.life - p.age) / 24));
    c.save();
    c.globalAlpha = fade;
    c.translate(p.x, p.y);
    c.rotate(p.angle);
    // Der schmaler werdende Streifen lässt das Papier flattern.
    c.scale(Math.cos(p.angle * 1.6), 1);
    c.fillStyle = p.color;
    c.fillRect(-p.w / 2, -p.h / 2, p.w, p.h);
    c.restore();
  }

  particles = particles.filter((p) => p.age < p.life && p.y < window.innerHeight + 40);
  if (particles.length) frame = requestAnimationFrame(tick);
  else teardown();
}

/** Fliegt gerade noch Papier? Verhindert, dass sich Salven stapeln. */
export function confettiActive(): boolean {
  return particles.length > 0;
}

/** Ein Schwung Konfetti. Mehrere Aufrufe kurz nacheinander addieren sich. */
export function confetti(opts: BurstOptions = {}): void {
  if (reducedMotion()) return;
  const c = ensureCanvas();
  if (!c) return;

  const {
    x = 0.5,
    y = 0.55,
    count = 70,
    colors = palette(),
    power = 13,
    spread = Math.PI * 0.75,
  } = opts;

  // Leere Farbliste (z.B. ein Abzeichen ohne Farbe) fällt auf die Palette zurück.
  const paper = (colors.length ? colors : DEFAULT_VARS.map((v) => `var(${v})`)).map(resolve);
  const originX = x * window.innerWidth;
  const originY = y * window.innerHeight;

  for (let i = 0; i < count; i++) {
    // Nach oben streuen: -90° ± spread/2.
    const angle = -Math.PI / 2 + (Math.random() - 0.5) * spread;
    const speed = power * (0.55 + Math.random() * 0.75);
    particles.push({
      x: originX + (Math.random() - 0.5) * 24,
      y: originY + (Math.random() - 0.5) * 12,
      vx: Math.cos(angle) * speed,
      vy: Math.sin(angle) * speed,
      w: 5 + Math.random() * 6,
      h: 8 + Math.random() * 6,
      color: paper[Math.floor(Math.random() * paper.length)],
      spin: (Math.random() - 0.5) * 0.34,
      angle: Math.random() * Math.PI,
      life: 90 + Math.random() * 60,
      age: 0,
    });
  }

  if (!frame) frame = requestAnimationFrame(tick);
}

/** Konfetti aus der Mitte eines Elements — z.B. dem gerade gedrückten Knopf. */
export function confettiFrom(el: Element | null, opts: BurstOptions = {}): void {
  if (!el) return confetti(opts);
  const r = el.getBoundingClientRect();
  confetti({
    ...opts,
    x: (r.left + r.width / 2) / window.innerWidth,
    y: (r.top + r.height / 2) / window.innerHeight,
  });
}

/**
 * Der große Auftritt: zwei Salven von links und rechts, dazu eine aus der
 * Mitte. Für Level-Ups und erreichte Wochenziele.
 */
export function confettiCheer(colors?: string[]): void {
  if (reducedMotion()) return;
  confetti({ x: 0.5, y: 0.42, count: 90, colors, power: 15 });
  window.setTimeout(() => confetti({ x: 0.08, y: 0.75, count: 45, colors, power: 17, spread: Math.PI / 2.4 }), 130);
  window.setTimeout(() => confetti({ x: 0.92, y: 0.75, count: 45, colors, power: 17, spread: Math.PI / 2.4 }), 230);
}

File: app/lib/widgets/confetti.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

import '../theme.dart';

/// Konfetti als Overlay: ein paar Dutzend Papierschnipsel mit Schwerkraft, die
/// sich nach dem Fall selbst wieder abräumen. Gegenstück zu
/// web/src/lib/confetti.ts — gleiche Physik, gleiche Farben.
///
/// Wer im System „Bewegung reduzieren“ gesetzt hat, bekommt gar nichts: die
/// Feier ist Zugabe, nie Information.

const _defaultColors = [
  C.tape,
  C.gradeYellow,
  C.gradeGreen,
  C.chalk,
  C.gradePurple,
];

const _gravity = 0.32;
const _drag = 0.987;

class _Particle {
  _Particle({
    required this.position,
    required this.velocity,
    required this.size,
    required this.color,
    required this.spin,
    required this.angle,
    required this.life,
  });

  Offset position;
  Offset velocity;
  final Size size;
  final Color color;
  final double spin;
  double angle;

  /// Lebensdauer in Frames.
  final double life;
  double age = 0;
}

final _random = Random();

/// Wie viele Konfetti-Schichten gerade fliegen.
int _live = 0;

/// Fliegt gerade noch Papier? Verhindert, dass sich Salven stapeln.
bool confettiActive() => _live > 0;

/// Ein Schwung Konfetti, ausgehend von [origin] (global, ohne Angabe die Mitte
/// des Bildschirms auf halber Höhe).
void burstConfetti(
  BuildContext context, {
  Offset? origin,
  List<Color> colors = _defaultColors,
  int count = 70,
  double power = 13,
  double spread = pi * 0.75,
}) {
  if (MediaQuery.maybeOf(context)?.disableAnimations ?? false) return;
  final overlay = Overlay.maybeOf(context, rootOverlay: true);
  if (overlay == null) return;

  final size = MediaQuery.of(context).size;
  final from = origin ?? Offset(size.width / 2, size.height * 0.55);
  final paper = colors.isEmpty ? _defaultColors : colors;

  final particles = [
    for (var i = 0; i < count; i++)
      () {
        // Nach oben streuen: -90° ± spread/2.
        final angle = -pi / 2 + (_random.nextDouble() - 0.5) * spread;
        final speed = power * (0.55 + _random.nextDouble() * 0.75);
        return _Particle(
          position: from +
              Offset(
                (_random.nextDouble() - 0.5) * 24,
                (_random.nextDouble() - 0.5) * 12,
              ),
          velocity: Offset(cos(angle) * speed, sin(angle) * speed),
          size: Size(5 + _random.nextDouble() * 6, 8 + _random.nextDouble() * 6),
          color: paper[_random.nextInt(paper.length)],
          spin: (_random.nextDouble() - 0.5) * 0.34,
          angle: _random.nextDouble() * pi,
          life: 90 + _random.nextDouble() * 60,
        );
      }(),
  ];

  late final OverlayEntry entry;
  _live++;
  entry = OverlayEntry(
    builder: (_) => IgnorePointer(
      child: _ConfettiLayer(
        particles: particles,
        onDone: () {
          _live--;
          entry.remove();
        },
      ),
    ),
  );
  overlay.insert(entry);
}

/// Der große Auftritt: eine Salve aus der Mitte, zwei von den Seiten. Für
/// Level-Ups und erreichte Wochenziele.
void cheerConfetti(BuildContext context, {List<Color> colors = _defaultColors}) {
  if (MediaQuery.maybeOf(context)?.disableAnimations ?? false) return;
  final size = MediaQuery.of(context).size;
  burstConfetti(
    context,
    origin: Offset(size.width / 2, size.height * 0.42),
    colors: colors,
    count: 90,
    power: 15,
  );
  for (final (delay, x, y) in [(130, 0.08, 0.75), (230, 0.92, 0.75)]) {
    Future<void>.delayed(Duration(milliseconds: delay), () {
      if (!context.mounted) return;
      burstConfetti(
        context,
        origin: Offset(size.width * x, size.height * y),
        colors: colors,
        count: 45,
        power: 17,
        spread: pi / 2.4,
      );
    });
  }
}

class _ConfettiLayer extends StatefulWidget {
  const _ConfettiLayer({required this.particles, required this.onDone});

  final List<_Particle> particles;
  final VoidCallback onDone;

  @override
  State<_ConfettiLayer> createState() => _ConfettiLayerState();
}

class _ConfettiLayerState extends State<_ConfettiLayer>
    with SingleTickerProviderStateMixin {
  late final Ticker _ticker;
  bool _done = false;

  @override
  void initState() {
    super.initState();
    _ticker = createTicker(_step)..start();
  }

  void _step(Duration _) {
    final height = context.size?.height ?? 4000;
    for (final p in widget.particles) {
      p.age++;
      p.velocity = Offset(
        p.velocity.dx * _drag,
        (p.velocity.dy + _gravity) * _drag,
      );
      p.position += p.velocity;
      p.angle += p.spin;
    }
    widget.particles.removeWhere((p) => p.age >= p.life || p.position.dy > height + 40);
    if (widget.particles.isEmpty) {
      if (_done) return;
      _done = true;
      _ticker.stop();
      // Nicht mitten im Ticker-Callback das Overlay abbauen.
      WidgetsBinding.instance.addPostFrameCallback((_) => widget.onDone());
      return;
    }
    setState(() {});
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size.infinite,
      painter: _ConfettiPainter(widget.particles),
    );
  }
}

class _ConfettiPainter extends CustomPainter {
  _ConfettiPainter(this.particles);

  final List<_Particle> particles;

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint();
    for (final p in particles) {
      // Am Ende ausblenden, statt die Schnipsel hart verschwinden zu lassen.
      final fade = ((p.life - p.age) / 24).clamp(0.0, 1.0);
      canvas.save();
      canvas.translate(p.position.dx, p.position.dy);
      canvas.rotate(p.angle);
      // Der schmaler werdende Streifen lässt das Papier flattern.
      canvas.scale(cos(p.angle * 1.6), 1);
      paint.color = p.color.withValues(alpha: fade);
      canvas.drawRect(
        Rect.fromCenter(center: Offset.zero, width: p.size.width, height: p.size.height),
        paint,
      );
      canvas.restore();
    }
  }

  @override
  bool shouldRepaint(_ConfettiPainter oldDelegate) => true;
}
